core/rawdb/freezer.go

```diff
@@ -546,3 +547,106 @@ func (f *freezer) freezeRange(nfdb *nofreezedb, number, limit uint64) (hashes []
 
 	return hashes, err
 }
+
+// isLegacyFn takes a freezer entry and determines if it is in an
+// old format and hence needs to be converted.
+type isLegacyFn = func([]byte) (bool, error)
+
+// convertLegacyFn takes a raw freezer entry in an older format and
+// returns it in the new format.
+type convertLegacyFn = func([]byte) ([]byte, error)
+
+// MigrateTable processes the entries in a given table in sequence
+// converting them to a new format if they're of an old format.
+func (f *freezer) MigrateTable(kind string, isLegacy isLegacyFn, convert convertLegacyFn) error {
+	if f.readonly {
+		return errReadOnly
+	}
+	f.writeLock.Lock()
+	defer f.writeLock.Unlock()
+
+	table, ok := f.tables[kind]
+	if !ok {
+		return errUnknownTable
+	}
+	// forEach iterates every entry in the table serially and in order, calling `fn`
+	// with the item as argument. If `fn` returns an error the iteration stops
+	// and that error will be returned.
+	forEach := func(t *freezerTable, fn func(uint64, []byte) error) error {
+		items := atomic.LoadUint64(&t.items)
+		batchSize := uint64(1024)
+		maxBytes := uint64(1024 * 1024)
+		for i := uint64(0); i < items; {
+			if i+batchSize > items {
+				batchSize = items - i
+			}
+			data, err := t.RetrieveItems(i, batchSize, maxBytes)
+			if err != nil {
+				return err
+			}
+			for j, item := range data {
+				if err := fn(i+uint64(j), item); err != nil {
+					return err
+				}
+			}
+			i += uint64(len(data))
+		}
+		return nil
+	}
+	ancientsPath := filepath.Dir(table.index.Name())
+	// Set up new dir for the migrated table, the content of which
+	// we'll at the end move over to the ancients dir.
+	migrationPath := filepath.Join(ancientsPath, "migration")
+	newTable, err := NewFreezerTable(migrationPath, kind, FreezerNoSnappy[kind])
+	if err != nil {
+		return err
+	}
+	batch := newTable.newBatch()
+	var out []byte
+	// Iterate through entries and transform them
+	forEach(table, func(i uint64, blob []byte) error {
+		if i%500000 == 0 {
+			log.Info("Processing legacy elements", "number", i)
+		}
+		legacy, err := isLegacy(blob)
+		if err != nil {
+			return err
+		}
+		out = blob
+		if legacy {
+			out, err = convert(blob)
```

Comment:
We can simplify the logic by calling convert functions directly. Since `isLegacy` checking itself is also non-trivial which involves several rlp.decode calls.